### PR TITLE
planner, executor: eliminate extra columns introduced by OrderBy upon Union

### PR DESCRIPTION
**Pull request by AndrewDi**
### What problem does this PR solve? 
fix issue #8189 and issue #8199

### What is changed and how it works?
When a `SELECT` statement contains `UNION` and `ORDER BY` cause contains select statement,TiDB will generate extra columns, so I add `LogicalProjection` when building union LogicalPlan to eliminate generated extra columns.

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

### CODE REVIEW

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -738,6 +740,20 @@ func (b *PlanBuilder) buildUnion(union *ast.UnionStmt) (LogicalPlan, error) {
 			return nil, errors.Trace(err)
 		}
 	}
+
+	// Fix issue 8189
```
Comment (winoros) [inline planner/core/logical_plan_builder.go]:
```suggestion
	// Fix issue #8189 (https://github.com/pingcap/tidb/issues/8189).
```

File: planner/core/logical_plan_builder.go [hunk 2]
```diff
@@ -738,6 +740,20 @@ func (b *PlanBuilder) buildUnion(union *ast.UnionStmt) (LogicalPlan, error) {
...
+	// If there are some expr column generated from order by, generate Projection plan to remove extra expr columns.
```
Comment (winoros) [inline planner/core/logical_plan_builder.go]:
```suggestion
	// If there are extra expressions generated from `ORDER BY` clause, generate a `Projection` to remove them.
```

**Comment by XuHuaiyu** [comment]
@AndrewDi It seems that we should cherry-pick this to release-2.1 and release-2.0?

**Pull request merged**